### PR TITLE
Descriptor Matcher Interface

### LINKED ISSUE #121: OpenCV Matcher Wrapper (Brute Force)
- Wrapper for OpenCV's matcher modules.

- Required for front end

- Starting with the BruteForce matcher, second will be Flann based.

### PR DESCRIPTION
**Pull request by navganti**
Resolves #121. Preliminary descriptor matcher interface, with an implementation of the brute force matcher.

One point of discussion is the showMatches method...I'm not sure the best way to show and remove a window. Currently it's set with a cv:::waitKey() to remove it, but perhaps the user would want another method.

#### Pre-Merge Checklist:
- [x] Code is documented in doxygen format
- [x] Code has automated tests
- [x] Zero compiler warnings
- [x] Formatted with `clang-format`

### CODE REVIEW

File: wave_vision/tests/matcher_tests/brute_force_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+const auto TEST_CONFIG = "tests/config/matcher/brute_force_matcher.yaml";
+const auto TEST_IMAGE_1 = "tests/data/image_center.png";
+const auto TEST_IMAGE_2 = "tests/data/image_right.png";
+
+// Test fixture to load image data
+class BFTest : public testing::Test {
+ protected:
+    BFTest() {
+        initDetector();
```
Comment (leokoppel) [inline wave_vision/tests/matcher_tests/brute_force_tests.cpp]: Since your class definition lists `FASTDetector fast;`, it was already _initialized_ (`FASTDetector()` was called) before you _assign_ to it here.

Comment (leokoppel) [inline wave_vision/tests/matcher_tests/brute_force_tests.cpp]: What is the point of a test fixture that is only used in one test? (a bit confusing to have `BFTest` and `BFTests` by the way).

File: wave_vision/tests/matcher_tests/brute_force_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        initDescriptor();
+        initMatcher();
+        this->image_1 = cv::imread(TEST_IMAGE_1, cv::IMREAD_COLOR);
+        this->image_2 = cv::imread(TEST_IMAGE_2, cv::IMREAD_COLOR);
+        this->keypoints_1 = this->fast.detectFeatures(this->image_1);
+        this->keypoints_2 = this->fast.detectFeatures(this->image_2);
```
Comment (leokoppel) [inline wave_vision/tests/matcher_tests/brute_force_tests.cpp]: You could probably test these functions individually  outside of the "visual confirmation"

File: wave_vision/include/wave/vision/matcher/brute_force_matcher.hpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    /** Constructs a BruteForceMatcher using parameters found in the
+     *  linked .yaml file.
+     *
+     *  @param config_path is the path to a .yaml file, containing the desired
+     *  parameters for the BruteForceMatcher.
+     */
+    BruteForceMatcher(const std::string &config_path);
+
+    /** Destructor */
+    ~BruteForceMatcher() = default;
```
Comment (navganti) [inline wave_vision/include/wave/vision/matcher/brute_force_matcher.hpp]: Haha I did this because in my last PR, you mentioned to set the destructor as default...

File: wave_vision/include/wave/vision/matcher/descriptor_matcher.hpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+    /** Function to display matches between two images.
+     *
+     *  @param image_1 the first image.
+     *  @param keypoints_1 the keypoints detected in the first image.
+     *  @param image_2 the second image.
+     *  @param keypoints_2 the keypoints detected in the second image.
+     *  @param matches the matches found between the images.
+     */
+    void showMatches(cv::Mat image_1,
```
Comment (navganti) [inline wave_vision/include/wave/vision/matcher/descriptor_matcher.hpp]: This was more on recommendation from @chutsu, to incorporate viewing within the actual class. I didn't incorporate this yet for either the detector or descriptor. I can remove this.

File: wave_vision/tests/matcher_tests/brute_force_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+// Libwave Headers
+#include "wave/wave_test.hpp"
+#include "wave/vision/detector/fast_detector.hpp"
+#include "wave/vision/descriptor/brisk_descriptor.hpp"
+#include "wave/vision/matcher/brute_force_matcher.hpp"
+
+namespace wave {
+
+const auto TEST_CONFIG = "tests/config/matcher/brute_force_matcher.yaml";
+const auto TEST_IMAGE_1 = "tests/data/image_center.png";
```
Comment (navganti) [inline wave_vision/tests/matcher_tests/brute_force_tests.cpp]: Figured they'd be nice for matches :)

File: wave_vision/src/matcher/brute_force_matcher.cpp [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+    // Create cv::BFMatcher object with the desired parameters
+    this->brute_force_matcher =
+      cv::BFMatcher::create(config.norm_type, config.cross_check);
+
+    // Store configuration parameters within member struct
+    this->current_config = config;
+}
+
+BruteForceMatcher::BruteForceMatcher(const std::string &config_path) {
```
Comment (navganti) [inline wave_vision/src/matcher/brute_force_matcher.cpp]: Okay. Will fix here, and also make a separate PR for removing them within the detector and descriptor class.

File: wave_vision/tests/matcher_tests/brute_force_tests.cpp [hunk 7]
```diff
@@ -0,0 +1,137 @@
+// C++ Headers
+#include <algorithm>
```
Comment (navganti) [inline wave_vision/tests/matcher_tests/brute_force_tests.cpp]: Nope, thanks. Habit from that cpplint script I was running before...

File: wave_vision/include/wave/vision/matcher/brute_force_matcher.hpp (hunk 3 above)
Comment (leokoppel) [inline wave_vision/include/wave/vision/matcher/brute_force_matcher.hpp]: Yeah, I should have suggested you remove it. There are three subtly different possibilities
- `{}`
- `= default`
- don't declare anything

Comment (navganti) [inline wave_vision/include/wave/vision/matcher/brute_force_matcher.hpp]: Yeah, I had the first one before. I'll make it undeclared.

File: wave_vision/src/matcher/brute_force_matcher.cpp (hunk 6 above)
Comment (navganti) [inline wave_vision/src/matcher/brute_force_matcher.cpp]: Or, @leokoppel do you mind if we just push this as is, and then I fix them for all of the modules at once? This way master will have the same implementation at each time.

File: wave_vision/include/wave/vision/matcher/descriptor_matcher.hpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+ *  Internally, this class, and all derived classes are wrapping various
+ *  descriptor matchers implemented in OpenCV. Further reference on descriptor
+ *  matchers can be found [here][opencv_descriptor_matchers].
+ *
+ *  [opencv_descriptor_matchers]:
+ *  http://docs.opencv.org/trunk/db/d39/classcv_1_1DescriptorMatcher.html
+ */
+class DescriptorMatcher {
+ public:
+    virtual ~DescriptorMatcher() = default;
```
Comment (navganti) [inline wave_vision/include/wave/vision/matcher/descriptor_matcher.hpp]: No. Actually, because I've also deleted the destructor and the showMatches method, there's really zero point of inheritance here (the only method was the overridden matchDescriptors method). So I've deleted the entire descriptorMatcher class.

File: wave_vision/include/wave/vision/matcher/brute_force_matcher.hpp [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+    /** Function to match keypoints descriptors between two images using the
+     *  BruteForceMatcher.
+     *
+     *  @param descriptors_1 the descriptors extracted from the first image.
+     *  @param descriptors_2 the descriptors extracted from the second image.
+     *
+     *  @return vector containing the best matches.
+     */
+    std::vector<cv::DMatch> matchDescriptors(cv::Mat &descriptors_1,
```
Comment (navganti) [inline wave_vision/include/wave/vision/matcher/brute_force_matcher.hpp]: Done. No more need for override, const added.

File: wave_vision/tests/matcher_tests/brute_force_tests.cpp (hunk 7 above)
Comment (navganti) [inline wave_vision/tests/matcher_tests/brute_force_tests.cpp]: Done.

File: wave_vision/tests/matcher_tests/brute_force_tests.cpp (hunk 1 above)
Comment (navganti) [inline wave_vision/tests/matcher_tests/brute_force_tests.cpp]: removed. I should fix this on my detector and descriptor too

File: wave_vision/tests/matcher_tests/brute_force_tests.cpp (hunk 2 above)
Comment (navganti) [inline wave_vision/tests/matcher_tests/brute_force_tests.cpp]: I mean - this is just to make sure that the matches actually are being computed and that they make sense. Really testing this is just making sure OpenCV's modules work. Just want to make sure that I am in fact able to see proper matches between objects.

I guess I could check if the size of matches is greater than zero as well, but I feel like the visual check here is more representative. What do you suggest @leokoppel?

File: wave_vision/tests/matcher_tests/brute_force_tests.cpp (hunk 5 above)
Comment (navganti) [inline wave_vision/tests/matcher_tests/brute_force_tests.cpp]: I'll remove Lenna in a cleanup for the other two. There's a lot of good tips you've shown me for each one of these modules that I should go back and incorporate.

**Comment by navganti** [comment]
Closing #144, as I will be incorporating #145 into this PR. The outlier rejection components should be included as features within the matcher.

File: wave_vision/src/matcher/brute_force_matcher.cpp (hunk 6 above)
Comment (navganti) [inline wave_vision/src/matcher/brute_force_matcher.cpp]: Also @leokoppel, this new function would be best implemented as a non member function, right?

File: wave_vision/tests/matcher_tests/brute_force_tests.cpp [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+    // Test has been confirmed visually
+    cv::drawMatches(
+      image_1, keypoints_1, image_2, keypoints_2, matches, img_with_matches);
+
+    cv::imshow("matches", img_with_matches);
+
+    cv::waitKey(0);
+}
+}
```
Comment (Jebediah) [inline wave_vision/tests/matcher_tests/brute_force_tests.cpp]: busted no newline

Comment (navganti) [inline wave_vision/tests/matcher_tests/brute_force_tests.cpp]: fixed.

**Comment by navganti** [comment]
Okay @leokoppel and @Jebediah I am going to merge this in. Just a heads up - the outlier rejection has resulted in some additions to this interface and some refactoring.

Closes #121.
